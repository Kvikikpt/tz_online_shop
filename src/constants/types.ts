export const TYPES = {
  UserRepository: Symbol.for('ClearCodeRepository'),
  ShopItemRepository: Symbol.for('ShopItemRepository'),
  AuthService: Symbol.for('AuthService'),
  AuthCheckMiddleware: Symbol.for('AuthCheckMiddleware'),
  AdminCheckMiddleware: Symbol.for('AdminCheckMiddleware'),
  ShopItemService: Symbol.for('ShopItemService'),
  RedisService: Symbol.for('RedisService'),
  CacheManager: Symbol.for('CacheManager'),
};
