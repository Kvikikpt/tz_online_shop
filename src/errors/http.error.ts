import { InternalError } from './internal.error';
import { ErrorCodes } from '../types';

export class HttpError extends InternalError {
  constructor(message: string, status: number) {
    super(message, status, ErrorCodes.HttpError);
  }
}
