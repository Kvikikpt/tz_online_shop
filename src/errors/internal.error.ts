import { ErrorCodes, HttpStatusCode } from '../types';

export abstract class InternalError extends Error {
  protected constructor(
    message = 'Internal error',
    public readonly statusCode: number = HttpStatusCode.InternalServerError,
    public readonly code: number = ErrorCodes.UndefinedError
  ) {
    super(message);
    Error.captureStackTrace(this, InternalError);
  }
}
