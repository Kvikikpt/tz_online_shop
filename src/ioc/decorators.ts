import { fluentProvide } from 'inversify-binding-decorators';

export const provideSingleton = (identifier: string | symbol): any =>
  fluentProvide(identifier).inSingletonScope().done(true);
