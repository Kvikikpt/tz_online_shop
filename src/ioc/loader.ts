require('../repositories');
require('../services');
require('../controllers');
require('../middlewares');
require('../managers');
