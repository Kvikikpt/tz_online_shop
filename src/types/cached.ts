type StringifiedType = Date;

export type Cached<T> = T extends Array<infer N>
  ? Array<Cached<N>>
  : {
      [P in keyof T]: T[P] extends StringifiedType
        ? T[P] | string
        : T[P] extends Record<string, unknown> | unknown[]
        ? Cached<T[P]>
        : T[P];
    };
