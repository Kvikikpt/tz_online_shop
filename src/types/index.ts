export * from './http-status-code';
export * from './error-codes';
export * from './cached';
