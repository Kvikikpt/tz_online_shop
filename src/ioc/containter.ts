/* eslint-disable import/first,import/newline-after-import */
import 'reflect-metadata';
import { Container } from 'inversify';
import { autoProvide, provide, buildProviderModule } from 'inversify-binding-decorators';

export * from './decorators';

/**
 * Creates container that will contain all dependencies
 */
const container = new Container({ skipBaseClassChecks: true });

import './loader'; // load files with "require()"

container.load(buildProviderModule()); // load "@provide" services

export { container, autoProvide, provide };
