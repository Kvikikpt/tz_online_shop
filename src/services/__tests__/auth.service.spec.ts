import { AuthService } from '../auth.service';
import { container } from '../../ioc/containter';
import { UserFixture } from './__fixtures__';
import { HttpError } from '../../errors';
import { HttpStatusCode } from '../../types';

jest.useFakeTimers('modern');

describe('AuthService', () => {
  let service: AuthService;

  jest.setSystemTime(new Date(12312312312));
  beforeEach(() => {
    service = container.resolve<AuthService>(AuthService);
  });

  describe('generateUserToken', () => {
    it('successfully generates user token', () => {
      expect(
        // @ts-ignore
        typeof AuthService.generateUserToken(UserFixture) === 'string'
      ).toBeTruthy();
    });
  });

  describe('encryptUserPassword', () => {
    it('successfully encrypts user password', () => {
      process.env.SALT = 'test_salt';
      expect(
        // @ts-ignore
        AuthService.encryptUserPassword('somepassword')
      ).toMatchSnapshot();
    });
  });

  describe('createUserAndGetToken', () => {
    let createNewUser: jest.SpyInstance;

    beforeEach(() => {
      createNewUser = jest
        //@ts-ignore
        .spyOn(service.userRepository, 'createNewUser')
        .mockResolvedValueOnce(UserFixture);
    });

    it('creating a new user', async () => {
      //@ts-ignore
      AuthService.generateUserToken = jest
        //@ts-ignore
        .fn(AuthService.generateUserToken)
        .mockImplementationOnce(() => 'test_token_lul');
      expect(await service.createUserAndGetToken(UserFixture)).toEqual('test_token_lul');
      expect(createNewUser).toBeCalled();
      //@ts-ignore
      expect(AuthService.generateUserToken).toBeCalled();
    });
  });

  describe('getTokenByPassportAndName', () => {
    let getUserByPasswordAndUsername: jest.SpyInstance;

    beforeEach(() => {
      getUserByPasswordAndUsername = jest
        //@ts-ignore
        .spyOn(service.userRepository, 'getUserByPasswordAndUsername');
    });

    it('gets token by passport and name', async () => {
      getUserByPasswordAndUsername.mockResolvedValueOnce(UserFixture);
      //@ts-ignore
      AuthService.generateUserToken = jest
        //@ts-ignore
        .fn(AuthService.generateUserToken)
        .mockImplementationOnce(() => 'test_token_lul');
      expect(await service.getTokenByPassportAndName('somepass', 'somename')).toEqual(
        'test_token_lul'
      );
      expect(getUserByPasswordAndUsername).toBeCalled();
    });

    it('if user not found, throws error', async () => {
      getUserByPasswordAndUsername.mockResolvedValueOnce(undefined);
      await expect(
        service.getTokenByPassportAndName('somepass', 'somename')
      ).rejects.toThrowError(new HttpError('User is not found', HttpStatusCode.NotFound));
      expect(getUserByPasswordAndUsername).toBeCalled();
    });
  });

  describe('getUserByToken', () => {
    it('gets user from token successfully', async () => {
      expect(
        // @ts-ignore
        await service.getUserByToken(AuthService.generateUserToken(UserFixture))
      ).toMatchSnapshot();
    });

    it('throws error, if token is invalid', async () => {
      await expect(service.getUserByToken('some_swong_token')).rejects.toThrowError();
    });
  });
});
