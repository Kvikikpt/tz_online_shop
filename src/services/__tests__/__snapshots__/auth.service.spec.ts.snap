// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthService encryptUserPassword successfully encrypts user password 1`] = `"6682ddf49a53a7011ce7"`;

exports[`AuthService getUserByToken gets user from token successfully 1`] = `
Object {
  "createdAt": "1970-05-23T12:05:12.312Z",
  "id": 1,
  "isAdmin": false,
  "name": "asd",
  "password": "999831345adba09d7c5d",
  "salt": "somenanoid",
}
`;
